README.md
# PyCraft
Minecraft clone written in Python.

This game is my attempt to clone Minecraft in Python. It does have some bugs in
it, but it is fairly bug-free, the main bug being an issue with Pyglet that
causes the exclusive mouse not to work correctly.

This is a heavily modified version of Fogleman's Minecraft:
https://github.com/fogleman/Minecraft

Controls:
WASD to move, 
Space to jump, 
1 to cycle block, 
Leftclick to destroy block, Rightclick to place block, 
T to open command console, 
Tab to toggle flight, 
Escape to free the mouse, click window to lock it again, 
V to save world.


CHANGELOG.txt
0.1.0
-Initial Release

0.2.0
-More blocks and biomes, bug fixes

0.3.0
-Updated graphics engine and added more content

0.4.0
-Reformatted biomes and put them in their own module, fixed water levels in ocean and swamp

0.5.0
-Added world saving and loading ! ! !


blocks.py
from pyglet import image
from pyglet.graphics import TextureGroup

group1 = TextureGroup(image.load('textures\\blocks\\watertexture.png').get_texture())
group2 = TextureGroup(image.load('textures\\blocks\\watertexture - Copy.png').get_texture())
group3 = TextureGroup(image.load('textures\\blocks\\watertexture - Copy (2).png').get_texture())

def tex_coord(x, y, n=4):
    """ Return the bounding vertices of the texture square.
    """
    m = 1.0 / n
    dx = x * m
    dy = y * m
    return dx, dy, dx + m, dy, dx + m, dy + m, dx, dy + m


def tex_coords(top, bottom, front, left=None, right=None, back=None):
    """ Return a list of the texture squares for the top, bottom and side.
    """
    if not left:
        left = front
    if not right:
        right = left
    if not back:
        back = front
    top = tex_coord(*top)
    bottom = tex_coord(*bottom)
    front = tex_coord(*front)
    left = tex_coord(*left)
    right = tex_coord(*right)
    back = tex_coord(*back)
    result = []
    result.extend(top)
    result.extend(bottom)
    result.extend(front)
    result.extend(left)
    result.extend(right)
    result.extend(back)
    return result


class Block:
    texture = tex_coords((0, 1), (0, 1), (0, 1))
    group = group1
    name = 'Block'

class Block2:
    texture = tex_coords((0, 0), (0, 0), (0, 0))
    group = group2
    name = 'Block2'

class Block3:
    texture = tex_coords((0, 0), (0, 0), (0, 0))
    group = group3
    name = 'Block3'

class GrassBlock(Block):
    texture = tex_coords((1, 0), (0, 1), (0, 0))
    name = 'Grass Block'

class Dirt(Block):
    name = 'Dirt'

class Sand(Block):
    texture = tex_coords((1, 1), (1, 1), (1, 1))
    name = 'Sand'

class Brick(Block):
    texture = tex_coords((2, 0), (2, 0), (2, 0))
    name = 'Brick'

class Stone(Block):
    texture = tex_coords((2, 1), (2, 1), (2, 1))
    name = 'Stone'

class OakWood(Block):
    texture = tex_coords((3, 1), (3, 1), (3, 1))
    name = 'Oak Wood'

class OakLeaf(Block):
    texture = tex_coords((3, 0), (3, 0), (3, 0))
    name = 'Oak Leaf'

class Water(Block):
    texture = tex_coords((0, 2), (0, 2), (0, 2))
    name = 'Water'

class BrownClay(Block):
    texture = tex_coords((1, 2), (1, 2), (1, 2))
    name = 'Brown Clay'

class SnowBlock(Block):
    texture = tex_coords((2, 2), (2, 2), (2, 2))
    name = 'Snow Block'

class OakLog(Block):
    texture = tex_coords((3, 2), (3, 2), (3, 1))
    name = 'Oak Log'

class CoalOre(Block):
    texture = tex_coords((0, 3), (0, 3), (0, 3))
    name = 'Coal Ore'

class IronOre(Block):
    texture = tex_coords((1, 3), (1, 3), (1, 3))
    name = 'Iron Ore'

class DiamondOre(Block):
    texture = tex_coords((2, 3), (2, 3), (2, 3))
    name = 'Diamond Ore'

class OakPlanks(Block):
    texture = tex_coords((3, 3), (3, 3), (3, 3))
    name = 'Oak Planks'

class GoldOre(Block2):
    texture = tex_coords((0, 3), (0, 3), (0, 3))
    name = 'Gold Ore'

class LapisOre(Block2):
    texture = tex_coords((1, 3), (1, 3), (1, 3))
    name = 'Lapis Ore'

class RedstoneOre(Block2):
    texture = tex_coords((2, 3), (2, 3), (2, 3))
    name = 'Redstone Ore'

class BirchWood(Block2):
    texture = tex_coords((3, 1), (3, 1), (3, 1))
    name = 'Birch Wood'

class BirchLog(Block2):
    texture = tex_coords((3, 2), (3, 2), (3, 1))
    name = 'Birch Log'

class BirchPlanks(Block2):
    texture = tex_coords((3, 3), (3, 3), (3, 3))
    name = 'Birch Planks'

class BirchLeaf(Block):
    texture = tex_coords((3, 0), (3, 0), (3, 0))
    name = 'Birch Leaves'

class DarkOakLog(Block2):
    texture = tex_coords((0, 2), (0, 2), (0, 1))
    name = 'Dark Oak Log'

class DarkOakPlanks(Block2):
    texture = tex_coords((0, 0), (0, 0), (0, 0))
    name = 'Dark Oak Planks'

class DarkOakWood(Block2):
    texture = tex_coords((0, 1), (0, 1), (0, 1))
    name = 'Dark Oak Wood'

class DarkOakLeaf(Block):
    texture = tex_coords((3, 0), (3, 0), (3, 0))
    name = 'Dark Oak Leaf'

class JungleLog(Block2):
    texture = tex_coords((1, 2), (1, 2), (1, 1))
    name = 'Jungle Log'

class JungleWood(Block2):
    texture = tex_coords((1, 1), (1, 1), (1, 1))
    name = 'Jungle Wood'

class JunglePlanks(Block2):
    texture = tex_coords((1, 0), (1, 0), (1, 0))
    name = 'Jungle Planks'

class JungleLeaf(Block2):
    texture = tex_coords((3, 0), (3, 0), (3, 0))
    name = 'Jungle Leaf'

class SpruceWood(Block3):
    texture = tex_coords((3, 2), (3, 2), (3, 2))
    name = 'Spruce Wood'

class SpruceLog(Block3):
    texture = tex_coords((3, 2), (3, 2), (3, 1))
    name = 'Spruce Log'

class SprucePlanks(Block3):
    texture = tex_coords((3, 3), (3, 3), (3, 3))
    name = 'Spruce Planks'

class SpruceLeaf(Block):
    texture = tex_coords((3, 0), (3, 0), (3, 0))
    name = 'Spruce Leaf'

class AcaciaLog(Block2):
    texture = tex_coords((2, 0), (2, 0), (2, 2))
    name = 'Acacia Log'

class AcaciaWood(Block2):
    texture = tex_coords((2, 2), (2, 2), (2, 2))
    name = 'Acacia Wood'

class AcaciaPlanks(Block2):
    texture = tex_coords((2, 1), (2, 1), (2, 1))
    name = 'Acacia Planks'

class AcaciaLeaf(Block):
    texture = tex_coords((3, 0), (3, 0), (3, 0))
    name = 'Acacia Leaf'

class CraftTable(Block3):
    texture = tex_coords((1, 0), (3, 3), (0, 0))
    name = 'Crafting Table'

class Furnace(Block3):
    texture = tex_coords((2, 1), (2, 1), (2, 0), left = (1, 1), back=(1, 1))
    name = 'Furnace'

all_blocks = Block.__subclasses__() + Block2.__subclasses__() + Block3.__subclasses__()

CAVE_DIE = [Stone, BrownClay, SnowBlock, Dirt, GrassBlock, Sand, DiamondOre, GoldOre, RedstoneOre, IronOre, CoalOre, LapisOre]


entities.py
import pyglet
import random
import blocks
import graphics_engine as ge
from pyglet.graphics import TextureGroup
from pyglet import image

group1 = TextureGroup(image.load('textures\\mobs\\mobs1.png').get_texture())

def tex_coord(x, y, n=4):
    """ Return the bounding vertices of the texture square.
    """
    m = 1.0 / n
    dx = x * m
    dy = y * m
    return dx, dy, dx + m, dy, dx + m, dy + m, dx, dy + m

def get(x, y, nx, ny):
    return x, y, x + nx, y, x + nx, y + ny, x, y + ny

def tex_coor(ti, bi, fi, li, ri, bai):
    t=[]
    t.extend(get(ti[0], ti[1], ti[2], ti[3]))
    t.extend(get(bi[0], bi[1], bi[2], bi[3]))
    t.extend(get(fi[0], fi[1], fi[2], fi[3]))
    t.extend(get(li[0], li[1], li[2], li[3]))
    t.extend(get(ri[0], ri[1], ri[2], ri[3]))
    t.extend(get(bai[0], bai[1], bai[2], bai[3]))
    return t


def tex_coords(top, bottom, front, left=None, right=None, back=None):
    """ Return a list of the texture squares for the top, bottom and side.
    """
    if not left:
        left = front
    if not right:
        right = left
    if not back:
        back = front
    top = tex_coord(*top)
    bottom = tex_coord(*bottom)
    front = tex_coord(*front)
    left = tex_coord(*left)
    right = tex_coord(*right)
    back = tex_coord(*back)
    result = []
    result.extend(top)
    result.extend(bottom)
    result.extend(front)
    result.extend(left)
    result.extend(right)
    result.extend(back)
    return result

def tex_coords_advanced(top, bottom, front, left=None, right=None, back=None, topsize=8, bottomsize=8, frontsize=8, leftsize=8, rightsize=8, backsize=8):
    """ Return a list of the texture squares for the top, bottom and side.
    """
    if not left:
        left = front
    if not right:
        right = left
    if not back:
        back = front
    top = tex_coord(*top,n=topsize)
    bottom = tex_coord(*bottom,n=bottomsize)
    front = tex_coord(*front,n=frontsize)
    left = tex_coord(*left,n=leftsize)
    right = tex_coord(*right,n=rightsize)
    back = tex_coord(*back,n=backsize)
    result = []
    result.extend(top)
    result.extend(bottom)
    result.extend(front)
    result.extend(left)
    result.extend(right)
    result.extend(back)
    return result


class BaseEntity:
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z
        self.width = 0
        self.height = 0
        self.depth = 0
        self.texture = None
        self.texturegroup = None
        self.dx = self.dy = self.dz = 0
        self.move_speed = 70
        self.m = 70

    def get_size(self):
        return self.width, self.height, self.depth

    def get_pos(self):
        return self.x, self.y, self.z

    def get_texture(self):
        return self.texture

    def get_group(self):
        return self.texturegroup

    def get_data(self):
        return (self.get_texture(), self.get_group())

    def move(self, world):
        self.move_speed -= 1
        if self.move_speed <= 0:
            self.move_speed = self.m
            self.dx = random.choice([self.dx for i in range(10)] + [-1, 0, 1])
            self.dz = random.choice([self.dz for i in range(10)] + [-1, 0, 1])
            self.x += self.dx
            self.z += self.dz
        if (self.x, round(self.y), self.z) in world.keys():
            if world[(self.x, round(self.y), self.z)] != blocks.Water:
                self.dy = .1
            else:
                self.dy = 0
        else:
            if (self.x, round(self.y) - 1, self.z) in world.keys() and world[(self.x, round(self.y) - 1, self.z)] != blocks.Water:
                self.dy = 0
            else:
                self.dy = -.1
        self.y += self.dy

    def update(self, model):
        self.draw(model)

    def draw(self, model):
        pass


class Pig(BaseEntity):
    def __init__(self, x, y, z):
        super(Pig, self).__init__(x, y, z)
        self.texturehead = tex_coords_advanced((0, 0), (0, 0), (1, 0), back=(0, 0), left=(0, 0))
        self.texturebody = tex_coords_advanced((2, 0), (2, 0), (0, 0), back=(3.5, 0), left=(2, 0))
        self.texturefoot = tex_coords_advanced((0, 0), (0, 0), (0, 0))
        self.texturegroup = TextureGroup(image.load('textures\\mobs\\mobs1.png').get_texture())

    def draw(self, model):
        for x in [self.x + -.35, self.x + .35]:
            for z in [self.z + -.15, self.z + .35]:
                ge.hide_rec(model, (x, self.y - .1, z))
        ge.hide_rec(model, (self.x, self.y + 0.2, self.z + 0.4))
        ge.hide_rec(model, (self.x, self.y, self.z + 0.1))
        self.move(model.world)
        ge.show_rec(model, (self.x, self.y + 0.2, self.z + 0.4), (0.8, 0.8, 0.8), self.texturehead, self.texturegroup)
        ge.show_rec(model, (self.x, self.y, self.z + 0.1), (0.8, 0.6, 1.3), self.texturebody, self.texturegroup)
        for x in [self.x + -.35, self.x + .35]:
            for z in [self.z + -.15, self.z + .35]:
                ge.show_rec(model, (x, self.y - .1, z), (0.1, 0.2, 0.1), self.texturefoot, self.texturegroup)


class EntityManager:
    def __init__(self, model, player):
        pyglet.clock.schedule_interval(self.update, 20, model, player)

    def update(self, dt, model, player):
        model.entities.append(Pig(random.randint(int(player.position[0] + -32), int(player.position[0] + 32)), 80, random.randint(int(player.position[2] + -32), int(player.position[2] + 32))))


noise_gen.py
import random as rand
import math

class NoiseParameters:
    def __init__(self, octaves, amplitude, smoothness, roughness, heightOffset):
        self.octaves = octaves #How fast the world gains height
        self.amplitude = amplitude #How sharp hills are
        self.smoothness = smoothness #How smooth the hills are
        self.roughness = roughness #How rough the hills are
        self.heightOffset = heightOffset #The height offset of the world

class NoiseGen:
    def __init__(self, seed, biome):
        self.seed = seed
        self.update_biome(biome)
        
    def update_biome(self, biome):
        heightOffset = 60
        if type(biome) == bytes: biome = biome.decode('utf-8')
        if biome == 'mountains' or biome == 'wooded_hills':
            octaves = 8
            amplitude = 70
            smoothness = 400
            roughness = 0.3
        if biome == 'plains' or biome == 'desert' or biome == 'tundra' or biome == 'savanna':
            octaves = 6
            amplitude = 40
            smoothness = 500
            roughness = 0.3
        if biome == 'forest' or biome == 'jungle' or biome == 'dark_forest' or biome == 'birch_forest' or biome == 'taiga' or biome == 'swamp':
            octaves = 7
            amplitude = 50
            smoothness = 450
            roughness = 0.3
        if biome == 'ocean':
            octaves = 7
            amplitude = 50
            smoothness = 450
            roughness = 0.3
            heightOffset = 0
        if biome == 'deep_ocean':
            octaves = 7
            amplitude = 50
            smoothness = 437
            roughness = 0.3
            heightOffset = -5
        if biome == 'badlands':
            octaves = 7
            amplitude = 90
            smoothness = 350
            roughness = 0.3
        if biome == 'eroded_badlands':
            octaves = 8
            amplitude = 80
            smoothness = 300
            roughness = 0.3
        else:
            octaves = 7
            amplitude = 35
            smoothness = 550
            roughness = 0.3
        self.biome = biome
        self.noiseParams = NoiseParameters(
            octaves, amplitude, smoothness, roughness, heightOffset
        )

    def _getNoise2(self, n):
        n += self.seed
        n = (int(n) << 13) ^ int(n)
        newn = (n * (n * n * 60493 + 19990303) + 1376312589) & 0x7fffffff
        return 1.0 - (float(newn) / 1073741824.0)

    def _getNoise(self, x, z):
        return self._getNoise2(x + z * 57)

    def _lerp(self, a, b, z):
        mu2 = (1.0 - math.cos(z * 3.14)) / 2.0
        return (a * (1 - mu2) + b * mu2)

    def _noise(self, x, z):
        floorX = float(int(x))
        floorZ = float(int(z))

        s = 0.0,
        t = 0.0,
        u = 0.0,
        v = 0.0;#Integer declaration

        s = self._getNoise(floorX,      floorZ)
        t = self._getNoise(floorX + 1,  floorZ)
        u = self._getNoise(floorX,      floorZ + 1)
        v = self._getNoise(floorX + 1,  floorZ + 1)

        rec1 = self._lerp(s, t, x - floorX)
        rec2 = self._lerp(u, v, x - floorX)
        rec3 = self._lerp(rec1, rec2, z - floorZ)
        return rec3

    def getHeight(self, x, z):
        totalValue = 0.0

        for a in range(self.noiseParams.octaves - 1):
            freq = math.pow(2.0, a)
            amp  = math.pow(self.noiseParams.roughness, a)
            totalValue += self._noise(
                (float(x)) * freq / self.noiseParams.smoothness,
                (float(z)) * freq / self.noiseParams.smoothness
            ) * self.noiseParams.amplitude

        result = (((totalValue / 2.1) + 1.2) * self.noiseParams.amplitude) + self.noiseParams.heightOffset

        final = (totalValue / 5) + self.noiseParams.heightOffset
        if final < 1: final = 1
        return final

    #def randIndex(self, x, z, lower, upper):
    #    o = self._lerp(upper, x, z) + self._getNoise(lower, z)
    #    s = self._getNoise(x, z)
    #    t = self._getNoise(z, x)
    #    u = self._getNoise2(upper)
    #    v = self._noise(s, x)
    #    w = self._lerp(t, o, v / 10)
    #    l = (w * self.seed) / (self._getNoise(t, u) * self.getHeight(lower, z))
    #    l /= self._noise(x * z, v * upper)
    #    l /= 124072573.27
    #    l = abs(l)
    #    if l == 0:
    #        l += abs(abs(self.randTempUpdate(98, 25) / upper) * self._getNoise(2, 1) + .5)
    #    l += lower
    #    if l < lower: l = lower
    #    if l > upper: l = upper
    #    return int(l)

    #def randTempUpdate(self, x, z):
    #    return self._getNoise(x, z) / 2.3
    def randTempUpdate(self, x, z):
        return rand.uniform(-0.2, 0.2)
    def randIndex(self, x, z, lower, upper):
        return rand.randint(lower, upper)

randIndex = NoiseGen(1010101010, 'mountains').randIndex
randTemp = NoiseGen(1010101010, 'mountains').randTempUpdate


graphics_engine.py
from pyglet.gl import *

__all__ = ["show_block", "hide_block", "show_sector", "hide_sector", "show_entity", "hide_entity"]

def cube_vertices(x, y, z, n):
    """ Return the vertices of the cube at position x, y, z with size 2*n.
    """
    return [
        x-n,y+n,z-n, x-n,y+n,z+n, x+n,y+n,z+n, x+n,y+n,z-n,  # top
        x-n,y-n,z-n, x+n,y-n,z-n, x+n,y-n,z+n, x-n,y-n,z+n,  # bottom
        x-n,y-n,z-n, x-n,y-n,z+n, x-n,y+n,z+n, x-n,y+n,z-n,  # left
        x+n,y-n,z+n, x+n,y-n,z-n, x+n,y+n,z-n, x+n,y+n,z+n,  # right
        x-n,y-n,z+n, x+n,y-n,z+n, x+n,y+n,z+n, x-n,y+n,z+n,  # front
        x+n,y-n,z-n, x-n,y-n,z-n, x-n,y+n,z-n, x+n,y+n,z-n,  # back
        ]

def rect_vertices(x, y, z, a, b, c):
    """ Return the vertices of the rectangle at position x, y, z with x-width a*2, z-depth b*2, and y-height c*2.
"""
    return [
        x-a,y+c,z-b, x-a,y+c,z+b, x+a,y+c,z+b, x+a,y+c,z-b,  # top
        x-a,y-c,z-b, x-a,y-c,z+b, x+a,y-c,z+b, x+a,y-c,z-b,  # bottom
        x-a,y-c,z-b, x-a,y-c,z+b, x-a,y+c,z+b, x-a,y+c,z-b,  # left
        x+a,y-c,z+b, x+a,y-c,z-b, x+a,y+c,z-b, x+a,y+c,z+b,  # right
        x-a,y-c,z+b, x+a,y-c,z+b, x+a,y+c,z+b, x-a,y+c,z+b,  # front
        x+a,y-c,z-b, x-a,y-c,z-b, x-a,y+c,z-b, x+a,y+c,z-b,  # back
        ]

class GraphicsEngine:
    def __init__(self):
        pass
    
    def show_block(self, model, position, immediate=True):
        """ Show the block at the given `position`. This method assumes the
        block has already been added with add_block()
        Parameters
        ----------
        position : tuple of len 3
            The (x, y, z) position of the block to show.
        immediate : bool
            Whether or not to show the block immediately.
        """
        texture = model.world[position]
        model.shown[position] = texture
        if immediate:
            self._show_block(model, position, texture)
        return texture

    def _show_block(self, model, position, texture):
        """ Private implementation of the `show_block()` method.
        Parameters
        ----------
        position : tuple of len 3
            The (x, y, z) position of the block to show.
        texture : list of len 3
            The coordinates of the texture squares. Use `tex_coords()` to
            generate.
        """
        x, y, z = position
        vertex_data = cube_vertices(x, y, z, 0.5)
        texture_data = list(texture.texture)
        # create vertex list
        # FIXME Maybe `add_indexed()` should be used instead
        model._shown[position] = model.batch.add(24, GL_QUADS, texture.group,
            ('v3f/static', vertex_data),
            ('t2f/static', texture_data))

    def hide_block(self, model, position, immediate=True):
        """ Hide the block at the given `position`. Hiding does not remove the
        block from the world.
        Parameters
        ----------
        position : tuple of len 3
            The (x, y, z) position of the block to hide.
        immediate : bool
            Whether or not to immediately remove the block from the canvas.
        """
        if position in model.shown.keys():
            model.shown.pop(position)
        if immediate:
            self._hide_block(model, position)

    def _hide_block(self, model, position):
        """ Private implementation of the 'hide_block()` method.
        """
        if position in model._shown.keys():
            model._shown.pop(position).delete()

    def show_sector(self, model, sector):
        """ Ensure all blocks in the given sector that should be shown are
        drawn to the canvas.
        """
        for position in model.sectors.get(sector, []):
            if position not in model.shown and model.exposed(position):
                model.show_block(position, False)

    def hide_sector(self, model, sector):
        """ Ensure all blocks in the given sector that should be hidden are
        removed from the canvas.
        """
        for position in model.sectors.get(sector, []):
            if position in model.shown:
                model.hide_block(position, False)

    def show_rec(self, model, position, size, texture, group, immediate=True):
        model.shown[position] = texture
        if immediate:
            self._show_rec(model, position, size, texture, group)

    def _show_rec(self, model, position, size, texture, group):
        x, y, z = position
        x1, y1, z1 = size
        vertex_data = rect_vertices(x, y, z, x1/2, z1/2, y1/2)
        texture_data = list(texture)
        # create vertex list
        # FIXME Maybe `add_indexed()` should be used instead
        model._shown[position] = model.batch.add(24, GL_QUADS, group,
            ('v3f/static', vertex_data),
            ('t2f/static', texture_data))

    def hide_rec(self, model, position, immediate=True):
        if position in model.shown.keys():
            model.shown.pop(position)
        if immediate:
            self._hide_rec(model, position)

    def _hide_rec(self, model, position):
        if position in model._shown.keys():
            model._shown.pop(position).delete()

graphics_engine = GraphicsEngine()

show_block = graphics_engine.show_block
hide_block = graphics_engine.hide_block
show_sector = graphics_engine.show_sector
hide_sector = graphics_engine.hide_sector
show_rec = graphics_engine.show_rec
hide_rec = graphics_engine.hide_rec


biomes.py
from blocks import *
import sys

DIRT = Dirt
GRASS = GrassBlock
LOG = OakLog
WOOD = OakWood
LEAF = OakLeaf
WATER = Water
SAND = Sand
STONE = Stone
BRICK = Brick
CLAY = BrownClay
SNOW = SnowBlock
PLANKS = OakPlanks

#biomes info

MOUNTAINS = {'structures': ['house',
                            'tower',
                            'tree'],
            'treeSpecs': {
                'woodType': ['oak',
                            'oak',
                            'birch'],
                'canGrowGiant': []
            },
            'surfBlocks': {
                'top': GRASS,
                'middle': DIRT,
                'bottom': STONE
            },
            'houseChance': 2000,
            'towerChance': 3000,
            'treeChance': 1000,
            'caveChance': 2000 + 50000,
            'requiredWaterHeight': 25,
            'caveSpecs': {
                'len': (14, 33),
                'width': (5, 10)
            }
}

PLAINS = {'structures': ['house',
                        'tower',
                        'tree'],
          'treeSpecs': {
              'woodType': ['oak',
                           'oak',
                          'oak',
                          'oak',
                          'birch'],
              'canGrowGiant': []
            },
            'surfBlocks': {
                    'top': GRASS,
                    'middle': DIRT,
                    'bottom': DIRT
            },
            'houseChance': 1000,
            'towerChance': 2000,
            'treeChance': 700,
            'caveChance': 7500 + 50000,
            'requiredWaterHeight': 25,
            'caveSpecs': {
                'len': (8, 17),
                'width': (3, 6)
            }
}

FOREST = {'structures': ['house',
                        'tower',
                        'tree'],
          'treeSpecs': {
              'woodType': ['oak',
                            'oak',
                            'oak',
                            'birch'],
             'canGrowGiant': []
            },
            'surfBlocks': {
                    'top': GRASS,
                    'middle': DIRT,
                    'bottom': DIRT
            },
            'houseChance': 1500,
            'towerChance': 3000,
            'treeChance': 155,
            'caveChance': 6000 + 50000,
            'requiredWaterHeight': 25,
            'caveSpecs': {
                'len': (8, 17),
                'width': (3, 6)
            }
}
OCEAN = {'structures': [],
        'surfBlocks': {
                'top': GRASS,
                'middle': DIRT,
                'bottom': DIRT
        },
        'caveChance': 10000 + 50000,
        'requiredWaterHeight': 25,
        'caveSpecs': {
            'len': (8, 17),
            'width': (3, 6)
        }
}
DEEP_OCEAN = {'structures': [],
            'surfBlocks': {
                    'top': GRASS,
                    'middle': DIRT,
                    'bottom': DIRT
            },
            'caveChance': 10000 + 50000,
            'requiredWaterHeight': 25,
            'caveSpecs': {
                'len': (8, 17),
                'width': (3, 6)
            }
}
WOODED_HILLS = {'structures': ['house',
                                'tower',
                                'tree'],
                'treeSpecs': {
                    'woodType': ['oak',
                                'oak',
                                'birch'],
                    'canGrowGiant': []
                },
                'surfBlocks': {
                        'top': GRASS,
                        'middle': DIRT,
                        'bottom': DIRT
                },
            'houseChance': 2000,
            'towerChance': 3000,
            'treeChance': 175,
            'caveChance': 2000 + 50000,
            'requiredWaterHeight': 25,
            'caveSpecs': {
                'len': (14, 33),
                'width': (5, 10)
            }
}
DESERT = {'structures': ['house'],
        'surfBlocks': {
                'top': SAND,
                'middle': SAND,
                'bottom': SAND
        },
        'houseChance': 2000,
        'caveChance': 6000 + 50000,
        'requiredWaterHeight': 25,
        'caveSpecs': {
            'len': (8, 17),
            'width': (3, 6)
        }
}
BADLANDS = {'structures': [],
            'surfBlocks': {
                    'top': CLAY,
                    'middle': CLAY,
                    'bottom': CLAY
            },
            'caveChance': 2000 + 50000,
            'requiredWaterHeight': 25,
            'caveSpecs': {
                'len': (14, 33),
                'width': (5, 10)
            }
}
JUNGLE = {'structures': ['house',
                        'tower',
                        'tree'],
        'treeSpecs': {
            'woodType': ['jungle',
                        'jungle',
                        'jungle',
                        'oak'],
            'giantTreeChance': 3,
            'canGrowGiant': ['jungle'],
            'giantHeight': (20, 30)
        },
        'surfBlocks': {
                'top': GRASS,
                'middle': DIRT,
                'bottom': DIRT
        },
        'houseChance': 3000,
        'towerChance': 2700,
        'treeChance': 115,
        'caveChance': 6000 + 50000,
        'requiredWaterHeight': 25,
        'caveSpecs': {
            'len': (8, 17),
            'width': (3, 6)
        }
}
ERODED_BADLANDS = {'structures': ['tower'],
                    'surfBlocks': {
                            'top': CLAY,
                            'middle': CLAY,
                            'bottom': CLAY
                    },
                    'towerChance': 3000,
                    'caveChance': 2000 + 50000,
                    'requiredWaterHeight': 50,
                    'caveSpecs': {
                        'len': (14, 33),
                        'width': (5, 14)
                    }
}
TUNDRA = {'structures': ['house',
                        'tower',
                        'tree'],
        'treeSpecs': {
            'woodType': ['spruce',
                        'spruce'],
            'canGrowGiant': []
        },
        'surfBlocks': {
                'top': SNOW,
                'middle': SNOW,
                'bottom': SNOW
        },
        'houseChance': 1000,
        'towerChance': 2000,
        'treeChance': 1000,
        'caveChance': 900 + 50000,
        'requiredWaterHeight': 25,
        'caveSpecs': {
            'len': (8, 17),
            'width': (3, 6)
        }
}
BIRCH_FOREST = {'structures': ['house',
                                'tower',
                                'tree'],
                'treeSpecs': {
                    'woodType': ['birch',
                                'birch'],
                    'giantTreeChance': 2,
                    'canGrowGiant': ['birch'],
                    'giantHeight': (10, 20)
                },
                'surfBlocks': {
                        'top': GRASS,
                        'middle': DIRT,
                        'bottom': DIRT
                },
                'houseChance': 1500,
                'towerChance': 3000,
                'treeChance': 155,
                'caveChance': 6000 + 50000,
                'requiredWaterHeight': 25,
                'caveSpecs': {
                    'len': (8, 17),
                    'width': (3, 6)
                }
}
DARK_FOREST = {'structures': ['house',
                            'tower',
                            'tree'],
                'treeSpecs': {
                    'woodType': ['dark oak',
                                'dark oak',
                                'dark oak',
                                'oak',
                                'birch'],
                    'giantTreeChance': 2,
                    'canGrowGiant': ['dark oak'],
                    'giantHeight': (3, 7)
                },
                'surfBlocks': {
                        'top': GRASS,
                        'middle': DIRT,
                        'bottom': DIRT
                },
                'houseChance': 1500,
                'towerChance': 3000,
                'treeChance': 115,
                'caveChance': 6000 + 50000,
                'requiredWaterHeight': 30,
                'caveSpecs': {
                    'len': (8, 17),
                    'width': (3, 6)
                }
}
TAIGA = {'structures': ['house',
                        'tower',
                        'tree'],
        'treeSpecs': {
            'woodType': ['spruce',
                        'spruce'],
            'canGrowGiant': []
        },
        'surfBlocks': {
                'top': GRASS,
                'middle': DIRT,
                'bottom': DIRT
        },
        'houseChance': 1500,
        'towerChance': 3000,
        'treeChance': 155,
        'caveChance': 6000 + 50000,
        'requiredWaterHeight': 25,
        'caveSpecs': {
            'len': (8, 17),
            'width': (3, 6)
        }
}
SAVANNA = {'structures': ['house',
                        'tower',
                        'tree'],
        'treeSpecs': {
            'woodType': ['acacia',
                        'acacia'],
            'canGrowGiant': []
        },
        'surfBlocks': {
                'top': GRASS,
                'middle': DIRT,
                'bottom': DIRT
        },
        'houseChance': 1000,
        'towerChance': 2000,
        'treeChance': 700,
        'caveChance': 7500 + 50000,
        'requiredWaterHeight': 25,
        'caveSpecs': {
            'len': (8, 17),
            'width': (3, 6)
        }
}
SWAMP = {'structures': ['house',
                        'tree'],
        'treeSpecs': {
            'woodType': ['oak'],
            'canGrowGiant': []
        },
        'surfBlocks': {
                'top': GRASS,
                'middle': DIRT,
                'bottom': DIRT
        },
        'houseChance': 2000,
        'towerChance': 2000,
        'treeChance': 900,
        'caveChance': 7500 + 50000,
        'requiredWaterHeight': 50,
        'caveSpecs': {
            'len': (8, 17),
            'width': (3, 6)
        }
}

TEMP = {
    'hot': [
            'desert',
            'savanna',
            'badlands',
            'badlands',
            'desert',
            'eroded_badlands',
            'desert'
    ],
    'warm': [
            'swamp',
            'swamp',
            'plains',
            'forest',
            'dark_forest',
            'birch_forest',
            'plains',
            'jungle',
            'dark_forest'
    ],
    'neutral': [
            'ocean',
            'ocean',
            'deep_ocean'
    ],
    'cool': [
            'taiga',
            'mountains',
            'wooded_hills',
            'mountains'
    ],
    'cold': [
            'tundra',
            'taiga'
    ]
}

for biome in [MOUNTAINS, PLAINS, FOREST, OCEAN, DEEP_OCEAN, WOODED_HILLS, DESERT, JUNGLE, BADLANDS, ERODED_BADLANDS, TUNDRA, BIRCH_FOREST, DARK_FOREST, TAIGA, SAVANNA, SWAMP]:
    biome['requiredWaterHeight'] = 60
